rust/noosphere-core/src/data/address.rs

```diff
@@ -1,91 +1,133 @@
-    pub identity: Did,
-    pub last_known_record: Option<Cid>,
+pub struct IdentityIpld {
+    pub did: Did,
+    pub link_record: Option<Link<LinkRecord>>,
 }
 
-impl AddressIpld {
-    /// If there is a last known record, attempt to retrieve it from storage as
-    /// a [Jwt] proof token
-    pub async fn get_proof<S: UcanJwtStore>(&self, store: &S) -> Option<Jwt> {
-        match &self.last_known_record {
+impl IdentityIpld {
+    /// If there is a [LinkRecord] for this [IdentityIpld], attempt to retrieve
+    /// it from storage
+    pub async fn link_record<S: UcanJwtStore>(&self, store: &S) -> Option<LinkRecord> {
+        match &self.link_record {
             Some(cid) => store
                 .read_token(cid)
                 .await
                 .unwrap_or(None)
-                .map(|jwt| jwt.into()),
+                .map(|jwt| LinkRecord(jwt.into())),
             _ => None,
         }
     }
+}
 
-    /// If a last known record is available, parses it as a [Ucan] and
-    /// looks for the referenced pointer to some data in IPFS (via a [Cid]
-    /// in the `fct` field).
-    pub async fn dereference<S: UcanJwtStore>(&self, store: &S) -> Option<Cid> {
-        match &self.last_known_record {
-            Some(cid) => {
-                let token = match store.require_token(cid).await {
-                    Ok(jwt) => jwt,
-                    Err(error) => {
-                        error!("Failed to look up token for {}: {}", cid, error);
-                        return None;
-                    }
-                };
-                let ucan = match Ucan::try_from(token) {
-                    Ok(ucan) => ucan,
-                    _ => return None,
-                };
-                let facts = ucan.facts();
-
-                for fact in facts {
-                    match fact.as_object() {
-                        Some(fields) => match fields.get("link") {
-                            Some(cid_string) => {
-                                match Cid::try_from(cid_string.as_str().unwrap_or_default()) {
-                                    Ok(cid) => return Some(cid),
-                                    Err(error) => {
-                                        warn!(
-                                            "Could not parse '{}' as name record link: {}",
-                                            cid_string, error
-                                        );
-                                        continue;
-                                    }
-                                }
-                            }
-                            None => {
-                                warn!("No 'link' field in fact, skipping...");
+/// A [LinkRecord] is a newtype that represents a JWT that ought to contain a
+/// [Cid] reference to a sphere
+#[derive(Debug, Eq, PartialEq, Clone, Serialize, Deserialize, Hash)]
+#[serde(from = "Jwt", into = "Jwt")]
+#[repr(transparent)]
+pub struct LinkRecord(Jwt);
+
+impl LinkRecord {
```

Comment:
As a "core type", wonder if `noosphere_ns::NsRecord` (or renamed to `noosphere_ns::LinkRecord`) could be replaced by this, or sharing some functionality.

Reply:
Yah, definitely possible!